File: python/src/wirebind/multiplex.py
from __future__ import annotations

from typing import Callable, Dict, Any, List

from .message import Message
from .encoding import Encoder
from .sender import Sender, RemoteSender

# Well-known channels
ROOT = 0


class Multiplexer:
    registry: Dict[int, Callable[[Message]]]
    send_function: Callable[[bytes], None]
    next_channel: int = 1
    encoder: Encoder
    remote_senders: List[RemoteSender]

    def __init__(self, send_function: Callable[[bytes], None]):
        self.registry = dict()
        self.encoder = Encoder(self)
        self.send_function = send_function
        self.remote_senders = []

    def set_root(self, root: Callable[[Message], None]):
        """Set the root callback for this Multiplexer."""
        if self.registry.get(0) is not None:
            raise ValueError("Root callback already set")
        self.registry[0] = root

    def send_root(self, message: Any):
        """Send a message to the remote root callback."""
        self.send(message, ROOT)

    def send(self, message: Any, channel: int):        
        message = Message(message, channel)
        message_enc = self.encoder.encode(message)
        self.send_function(message_enc)

    def register_sender(self, sender: Sender) -> int:
        """Register a Sender and return its Address."""
        channel = self.next_channel
        self.next_channel += 1
        self.registry[channel] = sender.send
        return channel

    def receive(self, message_bytes: bytes):
        """Receive an inbound message and route it to the appropriate callback."""
        message = self.encoder.decode(message_bytes)
        self.registry[message.channel](message.message)

    def cleanup(self):
        for remote_sender in self.remote_senders:
            remote_sender.destroy()
        # We need to clean up some fields manually because they may contain
        # circular references. If we don't do this, the Multiplexer may never
        # be garbage collected.
        self.remote_senders = None
        self.encoder = None
        self.registry = None

File: javascript/src/multiplex.ts
import { SenderEncoder } from "./encoder";
import { Sender } from "./sender";
import { Message } from "./types";

type ByteReceiver = (data: Uint8Array) => void;
export type MessageReceiver<T> = (data: T) => void;

// Well-known hosts
export const SERVER = 0;
export const CLIENT = 1;

// Well-known channels
export const ROOT = 0;

export class Multiplexer {
    registry: Map<number, MessageReceiver<any>>;
    sendFunction: ByteReceiver;
    nextChannel: number = 1;
    encoder: SenderEncoder;

    constructor(sendFunction: ByteReceiver) {
        this.registry = new Map();
        this.encoder = new SenderEncoder(this);

        this.sendFunction = sendFunction;
    }

    /** Set the root handler for messages from the remote party. */
    setRoot<T>(root: MessageReceiver<T>): void {
        if (this.registry.get(0) !== undefined) {
            throw new Error("Root callback already set");
        }
        this.registry.set(ROOT, root);
    }

    /** Send a message to the remote multiplexer's root handler. */
    sendRoot(message: any): void {
        this.send(message, ROOT);
    }

    /** Send a message to a channel on the remote multiplexer. */
    send(message: any, ch: number): void {
        const msg: Message<any> = {message, ch};
        console.log("Sending", msg)
        const messageEnc = this.encoder.encode(msg);
        this.sendFunction(messageEnc);
    }

    /** Register a local handler, returning a handle to the channel. */
    registerSender<T>(sender: Sender<T>): number {
        const channel = this.nextChannel;
        this.nextChannel += 1;
        this.registry.set(channel, sender.callback);
        return channel;
    }

    /** Receive a message from the remote multiplexer. */
    receive = (messageBytes: Uint8Array) => {
        const message = this.encoder.decode(messageBytes);

        if (this.registry.get(message.ch) === undefined) {
            throw new Error(`No callback registered for channel ${message.ch}`);
        }

        this.registry.get(message.ch)!(message.message);
    }
}
